libs/node-custom/src/cli/demo/index.ts

```diff
@@ -0,0 +1,99 @@
+import { Optional } from '@votingworks/types';
+import { Buffer } from 'buffer';
+import { writeFile } from 'fs/promises';
+import { Scanner } from '../../scanner';
+import { ReleaseType, ScanSide, SensorStatus } from '../../types';
+
+/**
+ * Simple demo program for using the Custom A4 scanner.
+ */
+export async function main(): Promise<number> {
+  let scanner: Optional<Scanner>;
+
+  process.on('SIGINT', async () => {
+    await scanner?.disconnect();
+    process.exit(0);
+  });
+
+  scanner = (await Scanner.open()).assertOk('failed to open scanner');
+
+  const model = (await scanner.getReleaseVersion(ReleaseType.Model)).assertOk(
+    'failed to get model'
+  );
+  const firmware = (
+    await scanner.getReleaseVersion(ReleaseType.Firmware)
+  ).assertOk('failed to get firmware');
+  const hardware = (
+    await scanner.getReleaseVersion(ReleaseType.Hardware)
+  ).assertOk('failed to get hardware');
+
+  console.log({ model, firmware, hardware });
+
+  const watcher = scanner.watchStatus();
+
+  process.stdout.write('> Waiting for paper to scan…\n');
+
+  for await (const result of watcher) {
+    if (result.isErr()) {
+      console.error(result.err());
+    } else {
+      const status = result.ok();
+
+      console.log(
+        `${status.isScanInProgress ? 'SCAN' : 'NO SCAN'}, ` +
+          `${status.isMotorOn ? 'MOVE' : 'NO MOVE'}, ` +
+          `LL ${SensorStatus[status.sensorInputLeftLeft]}, ` +
+          `CL ${SensorStatus[status.sensorInputCenterLeft]}, ` +
+          `CR ${SensorStatus[status.sensorInputCenterRight]}, ` +
+          `RR ${SensorStatus[status.sensorInputRightRight]}`
+      );
+
+      if (status.isTicketOnEnterA4 && !status.isMotorOn) {
+        watcher.stop();
+        console.log('Scanning…');
+        const scanResult = await scanner.scan(ScanSide.A_AND_B);
```

Comment:
I have some (probably buggy) locking in place that should prevent such race conditions. What should be happening, I think, is that the `Scanner#scan` promise will only resolve once the full scan is complete. There will be many underlying commands sent to the scanner in the process, but nothing callable from `Scanner`'s public API should be able to interfere with the scan in progress. So yes, the status does have `isScanInProgress` but the current API never allows that to be `true`.

Based on your experience with Plustek, is this desirable?

Reply:
I'm not sure. With Plustek, there were times where the `scan` command hung indefinitely and there was no way to interrupt it. That made me wish that all of the commands were synchronous/immediate and that I could control all of the status checking and error handling based on whatever physical effect happened as a result of the command.

That being said, since we have lower level control here, I think the abstractions that you're building could still give us that flexibility.

Thinking about the context in which this code will be used in our apps, I can say that it's useful to always be able to query for the current status of the scanner (or at least be notified on changes).

Some of this probably comes down to where we want the state-machine-esque logic to live. Is it in the driver or the application code? One lesson I took away from Plustek was that grouping as much of that together as possible was useful (whereas with most code I would instead try to build multiple layers of abstraction).

Random collection of thoughts, hope some of that is useful!

Reply:
`scan` will wait and block other calls until it's done. This isn't true of all the others: `move` does not wait until the move has finished. I'll make sure the JS Doc comments explain this.